layout: Make some major improvements to incremental layout to improve CNN.

pcwalton opened the pull request:
CNN is still too slow to be usable, but this is a partial solution.

r? @notriddle (feel free to reassign if you like)

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/13346)

<!-- Reviewable:end -->

pcwalton committed:
layout: Add a `REPOSITION` restyle damage type.

Separating out `REPOSITION` from `REPAINT` allows us to compute
stacking-context-relative positions without rebuilding the display list.
This saves a lot of time when responding to script-to-layout queries.

pcwalton committed:
layout: Remove reflow damage after reflowing replaced blocks.

pcwalton committed:
layout_thread: Don't build the display list for script queries unless
it's needed.

Makes cnn.com faster.

pcwalton committed:
layout: Remove reflow damage after reflowing table rows.

highfive commented:
Heads up! This PR modifies the following files:
 * @kichjang: components/script_layout_interface/restyle_damage.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify **unsafe code**. Please review it carefully!
* These commits modify layout code, but no tests are modified. Please consider adding a test!

notriddle commented on components/layout/sequential.rs:
```diff
@@ -78,7 +78,6 @@ pub fn build_display_list_for_subtree(flow_root: &mut Flow,
                                       root_stacking_context: &mut StackingContext,
                                       shared_layout_context: &SharedLayoutContext)
                                       -> Vec<DisplayItem> {
-    flow_root.traverse_preorder(&ComputeAbsolutePositions { layout_context: shared_layout_context });
```
Just to make sure I'm understanding this right, parallel does not and did not ever perform that traversal?

pcwalton replied:
Well, it did at one point, but I removed it because it was slower than sequential on all the Web sites I tried. Right now, this pass is always done sequentially.

notriddle replied:
Okay, that makes sense.

notriddle commented on components/layout_thread/lib.rs:
```diff
@@ -928,47 +932,56 @@ impl LayoutThread {
+                let display_list_needed = query_type.map(reflow_query_type_needs_display_list)
+                                                    .unwrap_or(false);
+                match (data.goal, display_list_needed) {
+                    (ReflowGoal::ForDisplay, _) | (ReflowGoal::ForScriptQuery, true) => {
```
Am I right in guessing that the `query_type` is `Some` if and only if `data.goal == ReflowGoal::ForScriptQuery`?

If so, either add an assertion for that `assert!(query_type.is_some() == data.goal == ReflowGoal::ForScriptQuery)`, or put the query type in the script reflow goal. The latter's more rustic, but might be complicated enough to save for another PR.

pcwalton replied:
Done.

notriddle reviewed (commented):
Review progress:
- [ ] 5a08e8b91031bfaafea666cd103648077b8cdf31
- [X] 28f141b2f5062ce5385a3305ea6a674bddc3434c
- [ ] b393ad3c2722bf37b51869dedd3f389e137cb6c9
- [X] 77fd27623b341e0a9b9943b4bc1257e777e453d5

notriddle reviewed (commented):
Review progress:
- [X] 5a08e8b
- [X] 28f141b
- [ ] b393ad3
- [X] 77fd276

pcwalton committed:
layout_thread: Add an assertion that reflows destined for the display
have no layout queries, and vice versa.

We should clean this up, but let's enforce it with an assert for now.

notriddle reviewed (approved):
Review finished:
- [X] 5a08e8b
- [X] 28f141b
- [X] b393ad3
- [X] 77fd276

As much as I would prefer if this had tests, we don't really have the infrastructure to test performance.

@bors-servo r+ (I wonder if homu will see this?)

notriddle commented:
Apparently not: @bors-servo r+

bors-servo commented:
:pushpin: Commit 0b1898e has been approved by `notriddle`

<!-- @bors-servo r=notriddle 0b1898ed8e4c739b5965f525eda208f89457d5ca -->

notriddle commented:
servo/homu#61

jdm commented:
@bors-servo: r-
This still fails tidy.

notriddle commented:
:facepalm:

pcwalton commented:
@bors-servo: r=notriddle

bors-servo commented:
:pushpin: Commit 0426ad3 has been approved by `notriddle`

<!-- @bors-servo r=notriddle 0426ad3b1b07bca9d4a99c947301967d13d36e6f -->

bors-servo commented:
:hourglass: Testing commit 0426ad3 with merge 5532162...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/3110)

highfive commented:
▶ FAIL [expected PASS] /css-backgrounds-3_dev/html4/background-repeat-repeat-y.htm
      └   → /css-backgrounds-3_dev/html4/background-repeat-repeat-y.htm 046f008dbdd5dbe1e80844169d96ba28ccb87526
    /css-backgrounds-3_dev/html4/reference/background-repeat-repeat-y.htm e6857ba68dc087e3588a61740846831c8fca2abb
    Testing 046f008dbdd5dbe1e80844169d96ba28ccb87526 == e6857ba68dc087e3588a61740846831c8fca2abb
    
      ▶ FAIL [expected PASS] /css-backgrounds-3_dev/html4/background-size-contain.htm
      └   → /css-backgrounds-3_dev/html4/background-size-contain.htm 40837300b75742f427a24d0f1e0678cac33a59f1
    /css-backgrounds-3_dev/html4/reference/background-size-contain.htm 369ed444761fa51e3bb4dad02090642f8c5dad5c
    Testing 40837300b75742f427a24d0f1e0678cac33a59f1 == 369ed444761fa51e3bb4dad02090642f8c5dad5c
    
      ▶ FAIL [expected PASS] /css-backgrounds-3_dev/html4/css3-background-origin-border-box.htm
      └   → /css-backgrounds-3_dev/html4/css3-background-origin-border-box.htm 7d8003864994c7fde43482d349f1f3a324f168f8
    /css-backgrounds-3_dev/html4/reference/css3-background-origin-border-box-ref.htm 6c41d68737ddcf038d8c3fd3eaded05408afda57
    Testing 7d8003864994c7fde43482d349f1f3a324f168f8 == 6c41d68737ddcf038d8c3fd3eaded05408afda57
    
      ▶ FAIL [expected PASS] /css-backgrounds-3_dev/html4/css3-background-origin-content-box.htm
      └   → /css-backgrounds-3_dev/html4/css3-background-origin-content-box.htm 5aaac5ae41497cdb437f5814f7896416c43116fb
    /css-backgrounds-3_dev/html4/reference/css3-background-origin-content-box-ref.htm facba575f917bbed941c7a80216fda61449bcb39
    Testing 5aaac5ae41497cdb437f5814f7896416c43116fb == facba575f917bbed941c7a80216fda61449bcb39
    
      ▶ FAIL [expected PASS] /css21_dev/html4/absolute-replaced-height-008.htm
      └   → /css21_dev/html4/absolute-replaced-height-008.htm fde77b15e0581f51cd1f3583b3df6061e414ba63
    /css21_dev/html4/reference/absolute-replaced-height-008-ref.htm d02649c040e9ed2ea8859060b0a7220500178559
    Testing fde77b15e0581f51cd1f3583b3df6061e414ba63 == d02649c040e9ed2ea8859060b0a7220500178559
    
      ▶ FAIL [expected PASS] /css21_dev/html4/absolute-replaced-height-009.htm
      └   → /css21_dev/html4/absolute-replaced-height-009.htm fde77b15e0581f51cd1f3583b3df6061e414ba63
    /css21_dev/html4/reference/absolute-replaced-height-008-ref.htm d02649c040e9ed2ea8859060b0a7220500178559
    Testing fde77b15e0</span><span class="stdout">581f51cd1f3583b3df6061e414ba63 == d02649c040e9ed2ea8859060b0a7220500178559
    
      ▶ FAIL [expected PASS] /css21_dev/html4/absolute-replaced-height-010.htm
      └   → /css21_dev/html4/absolute-replaced-height-010.htm 3bd46dd5a2c16115f7208926b611939bfd1f3e98
    /css21_dev/html4/reference/absolute-replaced-height-010-ref.htm dbf9d0b5c33353400db29c919e66cdebde7d44b5
    Testing 3bd46dd5a2c16115f7208926b611939bfd1f3e98 == dbf9d0b5c33353400db29c919e66cdebde7d44b5
    
      ▶ FAIL [expected PASS] /css21_dev/html4/absolute-replaced-height-016.htm
      └   → /css21_dev/html4/absolute-replaced-height-016.htm fde77b15e0581f51cd1f3583b3df6061e414ba63
    /css21_dev/html4/reference/absolute-replaced-height-008-ref.htm d02649c040e9ed2ea8859060b0a7220500178559
    Testing fde77b15e0581f51cd1f3583b3df6061e414ba63 == d02649c040e9ed2ea8859060b0a7220500178559
    
      ▶ FAIL [expected PASS] /css21_dev/html4/absolute-replaced-height-017.htm
      └   → /css21_dev/html4/absolute-replaced-height-017.htm 3bd46dd5a2c16115f7208926b611939bfd1f3e98
    /css21_dev/html4/reference/absolute-replaced-height-010-ref.htm dbf9d0b5c33353400db29c919e66cdebde7d44b5
    Testing 3bd46dd5a2c16115f7208926b611939bfd1f3e98 == dbf9d0b5c33353400db29c919e66cdebde7d44b5
    
      ▶ FAIL [expected PASS] /css21_dev/html4/absolute-replaced-height-022.htm
      └   → /css21_dev/html4/absolute-replaced-height-022.htm fde77b15e0581f51cd1f3583b3df6061e414ba63
    /css21_dev/html4/reference/absolute-replaced-height-008-ref.htm d02649c040e9ed2ea8859060b0a7220500178559
    Testing fde77b15e0581f51cd1f3583b3df6061e414ba63 == d02649c040e9ed2ea8859060b0a7220500178559
    
      ▶ FAIL [expected PASS] /css21_dev/html4/absolute-replaced-height-023.htm
      └   → /css21_dev/html4/absolute-replaced-height-023.htm fde77b15e0581f51cd1f3583b3df6061e414ba63
    /css21_dev/html4/reference/absolute-replaced-height-008-ref.htm d02649c040e9ed2ea8859060b0a7220500178559
    Testing fde77b15e0581f51cd1f3583b3df6061e414ba63 == d02649c040e9ed2ea8859060b0a7220500178559
    
      ▶ FAIL [expected PASS] /css21_dev/html4/absolute-replaced-height-024.htm
      └   → /css21_dev/html4/absolute-replaced-height-024.htm 3bd46dd5a2c16115f7208926b611939bfd1f3e98
    /css21_dev/html4/reference/absolute-replaced-height-010-ref.htm dbf9d0b5c33353400db29c919e66cdebde7d44b5
    Testing 3bd46dd5a2c16115f7208926b611939bfd1f3e98 == dbf9d0b5c33353400db29c919e66cdebde7d44b5
    
      ▶ FAIL [expected PASS] /css21_dev/html4/absolute-replaced-height-029.htm
      └   → /css21_dev/html4/absolute-replaced-height-029.htm fde77b15e0581f51cd1f3583b3df6061e414ba63
    /css21_dev/html4/reference/absolute-replaced-height-008-ref.htm d02649c040e9ed2ea8859060b0a7220500178559
    Testing fde77b15e0581f51cd1f3583b3df6061e414ba63 == d02649c040e9ed2ea8859060b0a7220500178559
    
      ▶ FAIL [expected PASS] /css21_dev/html4/absolute-replaced-height-030.htm
      └   → /css21_dev/html4/absolute-replaced-height-030.htm fde77b15e0581f51cd1f3583b3df6061e414ba63
    /css21_dev/html4/reference/absolute-replaced-height-008-ref.htm d02649c040e9ed2ea8859060b0a7220500178559
    Testing fde77b15e0581f51cd1f3583b3df6061e414ba63 == d02649c040e9ed2ea8859060b0a7220500178559
    
      ▶ FAIL [expected PASS] /css21_dev/html4/absolute-replaced-height-031.htm
      └   → /css21_dev/html4/absolute-replaced-height-031.htm 3bd46dd5a2c16115f7208926b611939bfd1f3e98
    /css21_dev/html4/reference/absolute-replaced-height-010-ref.htm dbf9d0b5c33353400db29c919e66cdebde7d44b5
    Testing 3bd46dd5a2c16115f7208926b611939bfd1f3e98 == dbf9d0b5c33353400db29c919e66cdebde7d44b5
    
      ▶ FAIL [expected PASS] /css21_dev/html4/abspos-026.htm
      └   → /css21_dev/html4/abspos-026.htm f1a62dc3f9391121c175d363c224fbfda63415db
    /css21_dev/html4/reference/abspos-025-ref.htm 8b4554e885d4a77c71d2bd6c20a072659f53c7e3
    Testing f1a62dc3f9391121c175d363c224fbfda63415db == 8b4554e885d4a77c71d2bd6c20a072659f53c7e3
    
      ▶ FAIL [expected PASS] /css21_dev/html4/abspos-025.htm
      └   → /css21_dev/html4/abspos-025.htm 997e38f3e2d69a73f9b180aa9c95a1121cde3323
    /css21_dev/html4/reference/abspos-025-ref.htm 8b4554e885d4a77c71d2bd6c20a072659f53c7e3
    Testing 997e38f3e2d69a73f9b180aa9c95a1121cde3323 == 8b4554e885d4a77c71d2bd6c20a072659f53c7e3
    
      ▶ FAIL [expected PASS] /css21_dev/html4/background-bg-pos-204.htm
      └   → /css21_dev/html4/background-bg-pos-204.htm de786a2bf1c39d17bac7ce9808b7d096d705aca2
    /css21_dev/html4/reference/background-bg-pos-204-ref.htm 2456ccc542e3de4c9d37f613e7fa57672dde2522
    Testing de786a2bf1c39d17bac7ce9808b7d096d705aca2 == 2456ccc542e3de4c9d37f613e7fa57672dde2522
    
      ▶ FAIL [expected PASS] /css21_dev/html4/background-bg-pos-206.htm
      └   → /css21_dev/html4/background-bg-pos-206.htm de786a2bf1c39d17bac7ce9808b7d096d705aca2
    /css21_dev/html4/reference/background-bg-pos-204-ref.htm 2456ccc542e3de4c9d37f613e7fa57672dde2522
    Testing de786a2bf1c39d17bac7ce9808b7d096d705aca2 == 2456ccc542e3de4c9d37f613e7fa57672dde2522
    
      ▶ FAIL [expected PASS] /css21_dev/html4/background-repeat-repeat-y.htm
      └   → /css21_dev/html4/background-repeat-repeat-y.htm 046f008dbdd5dbe1e80844169d96ba28ccb87526
    /css21_dev/html4/reference/background-repeat-repeat-y.htm e6857ba68dc087e3588a61740846831c8fca2abb
    Testing 046f008dbdd5dbe1e80844169d96ba28ccb87526 == e6857ba68dc087e3588a61740846831c8fca2abb

pcwalton commented:
@bors-servo: r=notriddle

The tests were failing because this patch set removed `REFLOW` and `REFLOW_OUT_OF_FLOW` damage bits from replaced absolutely-positioned blocks during block-size assignment, but this was too early as the final positions (and removal of those bits) happen during the `AbsoluteAssignBSizes` traversal instead.

bors-servo commented:
:pushpin: Commit 9f9f0d6 has been approved by `notriddle`

<!-- @bors-servo r=notriddle 9f9f0d604db05f779919b26a38515ccaee6b89cc -->

bors-servo commented:
:hourglass: Testing commit 9f9f0d6 with merge 01789e2...

bors-servo commented:
:broken_heart: Test failed - [linux-rel](http://build.servo.org/builders/linux-rel/builds/3334)

highfive commented:
▶ FAIL [expected PASS] /_mozilla/css/inline_block_opacity_change.html
      └   → /_mozilla/css/inline_block_opacity_change.html e80b50ad5eb4837811f922e16c65ed29d3152ed1
    /_mozilla/css/inline_block_opacity_change_ref.html febec1b4730afab195f280694dad47200b09ea3c
    Testing e80b50ad5eb4837811f922e16c65ed29d3152ed1 == febec1b4730afab195f280694dad47200b09ea3c

jdm commented:
@bors-servo: retry
* #13360

bors-servo commented:
:hourglass: Testing commit 9f9f0d6 with merge c6be14e...

bors-servo commented:
:broken_heart: Test failed - [linux-rel](http://build.servo.org/builders/linux-rel/builds/3342)

highfive commented:
▶ FAIL [expected PASS] /_mozilla/css/inline_block_opacity_change.html
      └   → /_mozilla/css/inline_block_opacity_change.html e80b50ad5eb4837811f922e16c65ed29d3152ed1
    /_mozilla/css/inline_block_opacity_change_ref.html febec1b4730afab195f280694dad47200b09ea3c
    Testing e80b50ad5eb4837811f922e16c65ed29d3152ed1 == febec1b4730afab195f280694dad47200b09ea3c

jdm commented:
@bors-servo: retry
#13360

bors-servo commented:
:zap: Previous build results for [arm32](http://build.servo.org/builders/arm32/builds/3095), [arm64](http://build.servo.org/builders/arm64/builds/3112), [linux-dev](http://build.servo.org/builders/linux-dev/builds/3167), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/3167), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/3162), [windows-dev](http://build.servo.org/builders/windows-dev/builds/1059) are reusable. Rebuilding only [linux-rel](http://build.servo.org/builders/linux-rel/builds/3342), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/3293)...

bors-servo commented:
:broken_heart: Test failed - [linux-rel](http://build.servo.org/builders/linux-rel/builds/3344)

highfive commented:
▶ FAIL [expected PASS] /_mozilla/css/inline_block_opacity_change.html
      └   → /_mozilla/css/inline_block_opacity_change.html e80b50ad5eb4837811f922e16c65ed29d3152ed1
    /_mozilla/css/inline_block_opacity_change_ref.html febec1b4730afab195f280694dad47200b09ea3c
    Testing e80b50ad5eb4837811f922e16c65ed29d3152ed1 == febec1b4730afab195f280694dad47200b09ea3c

jdm commented:
Looks like a permanent test failure to me.

notriddle commented:
An existing bug brought to the surface?

pcwalton commented:
@bors-servo: r=notriddle

The test was failing because properties that could possibly result in the formation of stacking contexts needed to be marked with `REPOSITION` (since they effectively change stacking-relative positions).

bors-servo commented:
:pushpin: Commit 031a4ed has been approved by `notriddle`

<!-- @bors-servo r=notriddle 031a4ed6ca7c50b7be708ff0bda7ff141f4a4c5a -->

bors-servo commented:
:hourglass: Testing commit 031a4ed with merge ad68314...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/3336)

notriddle commented:
@bors-servo retry

* #12654

bors-servo commented:
:hourglass: Testing commit 031a4ed with merge cbe5458...

bors-servo commented:
:sunny: Test successful - [arm32](http://build.servo.org/builders/arm32/builds/3148), [arm64](http://build.servo.org/builders/arm64/builds/3162), [linux-dev](http://build.servo.org/builders/linux-dev/builds/3232), [linux-rel](http://build.servo.org/builders/linux-rel/builds/3409), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/3216), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/3202), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/3341), [windows-dev](http://build.servo.org/builders/windows-dev/builds/1112)
